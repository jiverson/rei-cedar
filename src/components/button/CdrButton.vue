<template>
  <component
    :is="tag"
    :class="componentClass"
    :type="type"
  >
    <slot name="icon-left"/>
    <slot/>
    <slot name="icon-right"/>
  </component>
</template>

<!-- NOTE: REMOVED icon slot. use icon-left -->
<script lang="ts">
import { defineComponent, computed } from 'vue';
import clsx from 'clsx';
import fullWidth from '../../mixins/fullWidth';
import propValidator from '../../utils/propValidator';
import style from './styles/CdrButton.scss';

export default defineComponent({
  name: 'CdrButton',
  props: {
    /**
     * Controls render as button or anchor. {button, a}
     */
    tag: {
      type: String,
      default: 'button',
      validator: (value) => propValidator(value, ['button', 'a']),
    },
    /**
     * Sets the button type. {button, submit, reset}
     */
    type: {
      type: String,
      default: 'button',
      validator: (value) => propValidator(value, ['button', 'submit', 'reset']),
    },
    /**
     * Sets the button style
     */
    modifier: {
      type: String,
      default: 'primary',
      validator: (value) => propValidator(value, ['primary', 'secondary', 'sale', 'dark']),
    },
    size: {
      type: String,
      default: 'medium',
      validator: (value) => propValidator(
        value,
        ['small', 'medium', 'large'],
      ),
    },
    /**
     * Increases box-shadow around button to enhance contrast against background
     */
    elevated: {
      type: Boolean,
      default: false,
    },
    /**
     * Renders an icon-only button. Default slot is disabled. Overrides size and responsiveSize props.
     */
    iconOnly: {
      type: Boolean,
      default: false,
    },
    /**
     * Used in conjunction with `icon-only`, renders a border and background around the icon button
     */
    withBackground: {
      type: Boolean,
      default: false,
    },
  },
  // data() {
  //   return {
  //     style,
  //   };
  // },
  // computed: {

  //   iconClass() {
  //     const classes = [];
  //
  //     if ((this.$slots['icon-left'] || this.$slots.icon) && this.$slots.default) {
  //       /* only add class for buttons with text + icon on left */
  //       classes.push(this.modifyClassName(this.baseClass, 'has-icon-left'));
  //     }
  //
  //     if (this.$slots['icon-right'] && this.$slots.default) {
  //       /* only add class for buttons with text + icon on right */
  //       classes.push(this.modifyClassName(this.baseClass, 'has-icon-right'));
  //     }
  //
  //     if (this.elevated) {
  //       classes.push(this.modifyClassName(this.baseClass, 'elevated'));
  //     }
  //
  //     if (this.iconOnly) {
  //       classes.push(this.modifyClassName(this.baseClass, 'icon-only'));
  //     }
  //
  //     if (this.iconOnly && this.withBackground) {
  //       classes.push(this.modifyClassName(this.baseClass, 'with-background'));
  //     }
  //
  //     return classes.join(' ');
  //   },
  // },
  setup(props) {


    //   baseClass() {
    //     return 'cdr-button';
    //   },
    //   defaultClass() {
    //     return this.modifier ? undefined : this.modifyClassName(this.baseClass, 'primary');
    //   },
    //   buttonSizeClass() {
    //     return !this.iconOnly ? this.sizeClass : this.style[`cdr-button--icon-only-${this.size}`];
    //   },
    // \clsx(
    //   this.style[this.baseClass],
    const baseClass = style['cdr-button']
    const modifierClass = computed(() => );
    const buttonSizeClass = computed(() => ); // !!! SIZELCASS
    const fullWidthClass = computed(() => );
    const iconClass = computed(() => );
    // )
// TODO: can default/base/modifier be cleaned? seems like 1 too many!!!
    const componentClass = computed(() => clsx(baseClass, modifierClass, buttonSizeClass, fullWidthClass, iconClass));
    // TYPE


    return {
      componentClass,
      type: props.tag === 'button' ? props.type : null
    }
  }
});
</script>
