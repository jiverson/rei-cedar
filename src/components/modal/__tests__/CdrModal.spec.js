import { shallowMount, mount } from '@vue/test-utils';
import CdrModal from 'componentdir/modal/CdrModal';
import Vue from 'vue';
// import packageJson from '../../../../package.json';
import CdrButton from 'componentdir/button/CdrButton';

describe('CdrModal.vue', () => {
  it('default open, scrolling', async () => {
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        default: 'Sticky content',
      },
      computed: {
        scrolling: () => true,
      },
      attachToDocument: true,
    });

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('.cdr-modal__text-fade').exists()).toBe(true);

    wrapper.setProps({ opened: false });
    await wrapper.vm.$nextTick();

    setTimeout(() => {
      expect(wrapper.vm.reallyClosed).toBe(true);
      wrapper.destroy();
    }, 500);
  });

  it('leaves optional slots empty, handleOpened', async () => {
    const mockMeasureContent = jest.fn();
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: false,
        label: "Label is the modal title"
      },
      slots: {
        default: 'Main content',
      },
      methods: {
        measureContent: mockMeasureContent
      },
      attachToDocument: true,
    });

    wrapper.setProps({ opened: true });
    await wrapper.vm.$nextTick();
    expect(wrapper.element).toMatchSnapshot();
    
    setTimeout(() => {
      expect(mockMeasureContent).toHaveBeenCalled();
      wrapper.destroy();
    }, 300);
  });

  it('handleKeyDown', async () => {
    const mockMeasureContent = jest.fn();
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        default: 'Main content',
      },
      methods: {
        measureContent: mockMeasureContent,
      },
      attachToDocument: true,
    });

    wrapper.trigger('keydown', {
      key: 'a'
    });
    await wrapper.vm.$nextTick();

    wrapper.trigger('keydown', {
      key: 'Esc',
    });
    await wrapper.vm.$nextTick();

    wrapper.trigger('keydown', {
      key: 'Escape',
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().closed.length).toBe(2);
    wrapper.destroy();
  });

  it('scrolling and fullscreen snapshot', () => {
    const mockMeasureContent = jest.fn();
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      data() {
        return {
          offsetHeight: 400,
          scrollHeight: 500,
          fullscreen: true,
        };
      },
      methods: {
        measureContent: mockMeasureContent,
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      attachToDocument: true,
    });

    expect(wrapper.vm.scrolling).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
    wrapper.destroy();
  });

  it('removeNoScroll', () => {
    const mockMeasureContent = jest.fn();
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "My Modal Label",
      },
      methods: {
        measureContent: mockMeasureContent,
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      attachToDocument: true,
    });
    const { documentElement, body } = document;
    wrapper.vm.removeNoScroll();

    expect(documentElement.classList.contains('noscroll')).toBeFalsy();
    expect(body.classList.contains('noscroll')).toBeFalsy();

    wrapper.destroy();
  });

  it('handleFocus', () => {
    const mockMeasureContent = jest.fn();
    const wrapper = mount(CdrModal, {
      propsData: {
        opened: true,
        label: "My Modal Label",
      },
      methods: {
        measureContent: mockMeasureContent,
      },
      attachToDocument: true,
    });

    const button = wrapper.find('button').element;
    button.focus();
    expect(button).toBe(document.activeElement);

    wrapper.vm.handleFocus({ target: document.createElement('a') });
    expect(document.scrollTop).toBe(undefined);
    expect(document.scrollLeft).toBe(undefined);
    wrapper.destroy();
  });

  it('handleClosed', async () => {
    global.scrollTo = jest.fn();
    const spyMeasureContent = jest.fn();
    const spyHandleOpened = jest.fn();
    
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true, 
        label: "My Modal Label",
      },
      data() {
        return {
          offset: { x: 0, y: 0 }
        };
      },
      methods: {
        measureContent: spyMeasureContent,
        handleOpened: spyHandleOpened,
      },
      attachToDocument: true,
    });

    wrapper.vm.handleClosed();
    await wrapper.vm.$nextTick();

    setTimeout(() => {
      expect(wrapper.vm.reallyClosed).toBe(true);
      expect(wrapper.vm.unsubscribe).toBe(null);
    }, 500);
  });

  it('resize event', async () => {
    const spyMeasureContent = jest.fn();
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        default: 'Sticky content',
      },
      methods: {
        measureContent: spyMeasureContent,
      },
      attachToDocument: true,
    });

    window.dispatchEvent(new Event('resize'));
    await wrapper.vm.$nextTick();

    setTimeout(() => {
      expect(spyMeasureContent).toHaveBeenCalled();
      wrapper.destroy();
    }, 500);
  });
});