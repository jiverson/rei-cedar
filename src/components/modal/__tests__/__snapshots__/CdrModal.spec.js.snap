// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrModal.vue default open, scrolling 1`] = `
<div
  class="cdr-modal"
  role="presentation"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="cdr-modal__innerWrap"
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <cdr-text-stub
                  modifier="heading-serif-600"
                  space=""
                  tag="h1"
                >
                  Label is the modal title
                </cdr-text-stub>
              </div>
              <cdr-button-stub
                aria-label="Close"
                class="cdr-modal__close-button"
                icononly="true"
                id="close-modal-button"
                size="medium"
                space=""
                tag="button"
                type="button"
                withbackground="true"
              >
                <iconxlg-stub
                  class="cdr-button__icon"
                  inheritcolor="true"
                  size="medium"
                  space=""
                />
              </cdr-button-stub>
            </div>
            <div
              class="cdr-modal__text"
              role="document"
            >
              <div
                class="cdr-modal__text-content"
                style="max-height: -80px;"
                tabindex="0"
              >
                Sticky content
              </div>
              <div
                class="cdr-modal__text-fade"
              />
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`CdrModal.vue leaves optional slots empty, handleOpened 1`] = `
<div
  class="cdr-modal"
  role="presentation"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="cdr-modal__innerWrap"
        style=""
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <cdr-text-stub
                  modifier="heading-serif-600"
                  space=""
                  tag="h1"
                >
                  Label is the modal title
                </cdr-text-stub>
              </div>
              <cdr-button-stub
                aria-label="Close"
                class="cdr-modal__close-button"
                icononly="true"
                id="close-modal-button"
                size="medium"
                space=""
                tag="button"
                type="button"
                withbackground="true"
              >
                <iconxlg-stub
                  class="cdr-button__icon"
                  inheritcolor="true"
                  size="medium"
                  space=""
                />
              </cdr-button-stub>
            </div>
            <div
              class="cdr-modal__text"
              role="document"
            >
              <div
                class="cdr-modal__text-content"
                style="max-height: -80px;"
                tabindex="0"
              >
                Main content
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`CdrModal.vue scrolling and fullscreen snapshot 1`] = `
<div
  class="cdr-modal"
  role="presentation"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="cdr-modal__innerWrap"
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <cdr-text-stub
                  modifier="heading-serif-600"
                  space=""
                  tag="h1"
                >
                  Label is the modal title
                </cdr-text-stub>
              </div>
              <cdr-button-stub
                aria-label="Close"
                class="cdr-modal__close-button"
                icononly="true"
                id="close-modal-button"
                size="medium"
                space=""
                tag="button"
                type="button"
                withbackground="true"
              >
                <iconxlg-stub
                  class="cdr-button__icon"
                  inheritcolor="true"
                  size="medium"
                  space=""
                />
              </cdr-button-stub>
            </div>
            <div
              class="cdr-modal__text"
              role="document"
            >
              <div
                class="cdr-modal__text-content"
                style="max-height: -32px;"
                tabindex="0"
              />
              <div
                class="cdr-modal__text-fade"
              />
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;
